scripts/box/migrate_to_external_account.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Script to migrate Box credentials from user settings object to external
+account objects.
+
+Changes:
+ - Create external account for authorized user settings
+ - Attach external account to user settings
+ - Attach external account to all node settings
+"""
+
+import sys
+import logging
+from modularodm import Q
+from modularodm.storage.base import KeyExistsException
+
+from website.app import init_app
+from scripts import utils as script_utils
+from framework.transactions.context import TokuTransaction
+
+from website.addons.box.model import BoxUserSettings
+from website.addons.box.model import BoxNodeSettings
+from website.oauth.models import ExternalAccount
+
+logger = logging.getLogger(__name__)
+
+
+def do_migration(records):
+    database['boxnodesettings'].update({'user_settings': {'$type': 2}}, {'$rename': { 'user_settings': 'foreign_user_settings'}}, multi=True)
+
+    for user_addon in records:
+        user = user_addon.owner
+        old_account = user_addon.oauth_settings
+
+        logger.info('Record found for user {}'.format(user._id))
+
+        # Create/load external account and append to user
+        try:
+            account = ExternalAccount(
+                provider='box',
+                provider_name='Box',
+                display_name=old_account.username,
+                oauth_key=old_account.access_token,
+                refresh_token=old_account.refresh_token,
+                provider_id=old_account.user_id,
+                expires_at=old_account.expires_at,
+            )
+            account.save()
+        except KeyExistsException:
+            # ... or get the old one
+            account = ExternalAccount.find_one(
+                Q('provider', 'eq', 'box') &
+                Q('provider_id', 'eq', old_account.user_id)
+            )
+            assert account is not None
+        user.external_accounts.append(account)
+        user.save()
+
+        # Remove oauth_settings from user settings object
+        user_addon.oauth_settings = None
```

Comment:
It's probably worth noting that it breaks on the second iteration. That test was written to check what happens when multiple users have the same Box user, which is possible on production. It breaks because, after migrating the first user's authorization, that auth info is deleted before the second user is migrated.